UnitTests/GitUITests/UITest.cs

```diff
@@ -27,29 +29,41 @@ void HandleApplicationIdle(object sender, EventArgs e)
         }
 
         public static void RunForm<T>(
-            Action showDialog,
+            Action showForm,
             Func<T, Task> runAsync)
             where T : Form
         {
+            // Start runAsync before calling showForm.
+            // The latter might block until the form is closed, especially if using Application.Run(form).
+
             // Avoid using ThreadHelper.JoinableTaskFactory for the outermost operation because we don't want the task
             // tracked by its collection. Otherwise, test code would not be able to wait for pending operations to
             // complete.
             var test = ThreadHelper.JoinableTaskContext.Factory.RunAsync(async () =>
             {
-                await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
-                await WaitForIdleAsync();
-                var dialog = Application.OpenForms.OfType<T>().Single();
+                // Wait for the form to be opened by the test thread.
+                T form;
+                do
+                {
+                    await WaitForIdleAsync(); // does also SwitchToMainThreadAsync
```

Comment:
❗️  This isn't how asynchronous methods work. Even though `WaitForIdleAsync` calls `SwitchToMainThreadAsync`, there is no guarantee that a continuation will behave the same way. The preceding call to `SwitchToMainThreadAsync` *must not* be removed (correctness requirement). Multiple calls to this are not a problem; it will quickly NOP if the caller is already on the main thread.

Reply:
OK, I will restore it. I guess I understand why. That's an evil trap for async beginners like me.